book-content/chapters/11-annotations-and-assertions.md

```diff
@@ -0,0 +1,1196 @@
+
+In this example, we're creating an error by passing a string into a function that expects a number.
+
+```typescript
+function addOne(num: number) {
+  return num + 1;
+}
+
+// @ts-expect-error
+const result = addOne("one");
+```
+
+But the error doesn't show up in the editor, because we told TypeScript to expect it.
+
+However, if we pass a number into the function, the error will show up:
+
+```ts twoslash
+// @errors: 2578
+function addOne(num: number) {
+  return num + 1;
+}
+
+// ---cut---
+// @ts-expect-error
+const result = addOne(1);
+```
+
+So, TypeScript expects every `@ts-expect-error` directive to be _used_ - to be followed by an error.
+
+Frustratingly, `@ts-expect-error` doesn't let you expect a specific error, but only that an error will occur.
+
+### `@ts-ignore`
+
+The `@ts-ignore` directive behaves a bit differently than `@ts-expect-error`. Instead of _expecting_ an error, it _ignores_ any errors that do occur.
+
+Going back to our `addOne` example, we can use `@ts-ignore` to ignore the error that occurs when passing a string into the function:
+
+```typescript
+// @ts-ignore
+const result = addOne("one");
+```
+
+But if we later fix the error, `@ts-ignore` won't tell us that it's unused:
+
+```typescript
+// @ts-ignore
+const result = addOne(1); // No errors here!
+```
+
+In general, `@ts-expect-error` is more useful than `@ts-ignore`, because it tells you when you've fixed the error. This means you can get a warning to remove the directive.
+
+### `@ts-nocheck`
+
+Finally, The `@ts-nocheck` directive will completely remove type checking for a file.
+
+To use it, add the directive at the top of your file:
+
+```tsx
+// @ts-nocheck
+```
+
+With all checking disabled, TypeScript won't show you any errors, but it also won't be able to protect you from any runtime issues that might show up when you run your code.
+
+Generally speaking, you shouldn't use `@ts-nocheck`. I've personally lost hours of my life to working in large files where I didn't notice that `@ts-nocheck` was at the top.
+
+### Suppressing Errors Vs `as any`
+
+There's one tool in a TypeScript developers' toolkit that _also_ suppresses errors, but isn't a comment directive - `as any`.
+
+`as any` is an extremely powerful tool because it combines a lie to TypeScript (`as`) with a type that disables all type checking (`any`).
+
+This means that you can use it to suppress nearly any error. Our example above? No problem:
+
+```typescript
+const result = addOne({} as any);
+```
+
+`as any` turns the empty object into `any`, which disables all type checking. This means that `addOne` will happily accept it.
+
+#### `as any` vs Error Suppression Directives
+
+When there's a choice with how to suppress an error, I prefer using `as any`. Error suppression directives are too broad - they target the entire line of code. This can lead to accidentally suppressing errors that you didn't mean to:
```

Comment:
It's also somewhat "easy" to reason about `as any` because there are certain rules around it. We know how it propagates through types etc. A suppression might easily result in an error type that is displayed as `any` on hover... but it doesn't really have well-defined rules around it and there is no plan to codify them. All bets are off at that point. In fact, one behavior around them changed in the recent 5.5... so it actually broke some libraries/tools relying on the old behavior and there is no intention to bring back the old behavior